src/window/macos.rs

```diff
@@ -0,0 +1,152 @@
+use icrate::{
+    AppKit::{
+        NSColor, NSEvent, NSView, NSViewMinYMargin, NSViewWidthSizable, NSWindow,
+        NSWindowStyleMaskBorderless, NSWindowStyleMaskClosable,
+        NSWindowStyleMaskFullSizeContentView, NSWindowStyleMaskMiniaturizable,
+        NSWindowStyleMaskResizable, NSWindowStyleMaskTitled, NSWindowTitleHidden,
+    },
+    Foundation::{NSPoint, NSRect, NSSize},
+};
+use objc2::{declare_class, msg_send_id, mutability::InteriorMutable, rc::Id, ClassType};
+
+use csscolorparser::Color;
+use raw_window_handle::{HasRawWindowHandle, RawWindowHandle};
+use winit::window::Window;
+
+use crate::{cmd_line::CmdLineSettings, error_msg, frame::Frame, settings::SETTINGS};
+
+use super::WindowSettings;
+
+declare_class!(
+    // A view to simulate the double-click-to-zoom effect.
+    struct TitlebarClickHandler;
+
+    unsafe impl ClassType for TitlebarClickHandler {
+        type Super = NSView;
+        type Mutability = InteriorMutable;
+        const NAME: &'static str = "TitlebarClickHandler";
+    }
+
+    unsafe impl TitlebarClickHandler {
+        #[method(mouseDown:)]
+        unsafe fn mouse_down(&self, event: &NSEvent) {
+            if event.clickCount() == 2 {
+                self.window().unwrap().zoom(Some(&self));
+            }
+        }
+    }
+);
+
+impl TitlebarClickHandler {
+    pub fn new() -> Id<TitlebarClickHandler> {
+        unsafe { msg_send_id![Self::alloc(), init] }
+    }
+}
+
+lazy_static! {
+    static ref TITLEBAR_HEIGHT: f64 = MacosWindowFeature::calculate_titlebar_height();
+}
+
+// A trait for winit Window on macos, by which we can easily manipulate the window.
+pub struct MacosWindowFeature {
+    ns_window: Id<NSWindow>,
+    has_titlebar: bool,
+}
+
+impl MacosWindowFeature {
+    pub fn from_winit_window(window: &Window) -> MacosWindowFeature {
+        let ns_window = if let RawWindowHandle::AppKit(handle) = window.raw_window_handle() {
+            unsafe { Id::new(handle.ns_window as *mut NSWindow).unwrap() }
```

Comment:
The pointer does not have `+1` retain count, so this ends up doing a double-free.

```suggestion
            unsafe { Id::retain(handle.ns_window as *mut NSWindow).unwrap() }
```

See also https://github.com/madsmtm/objc2/issues/543.